### PR TITLE
sandbox/cgroup: do not be so eager to fail when paths do not exist

### PR DESCRIPTION
**Pull request by bboozzoo**
The groups that we walk through may be going away at any time. Try to be more
robust in such scenarios.

### CODE REVIEW

File: sandbox/cgroup/freezer.go [hunk 1]
```diff
@@ -140,13 +145,29 @@ func applyToSnap(snapName string, action func(groupName string) error, skipErrs
...
+	if err != nil {
+		return err
+	}
+	_, errW := f.Write(data)
+	errC := f.Close()
+	// pick the right error
+	if errC != nil && errW == nil {
+		return errC
+	}
+	return errW
```
Comment (stolowski) [inline sandbox/cgroup/freezer.go:168]: How about
```
if errW != nil {
    return errW
}
return errC
```
or am I missing something?

File: sandbox/cgroup/freezer.go [hunk 2]
```diff
@@ -140,13 +145,29 @@ func applyToSnap(snapName string, action func(groupName string) error, skipErrs
...
-		if err := action(name); err != nil && !skipErrs {
+		if err := action(name); err != nil && (skipError == nil || !skipError(err)) {
 			return err
 		}
 		return filepath.SkipDir
 	})
 }
 
+// writeFile can be used as a drop-in replacement for ioutil.WriteFile, but does
+// not create a file when does not exist
```
Comment (stolowski) [inline sandbox/cgroup/freezer.go:156]:
```suggestion
// not create a file when it does not exist
```
?

File: sandbox/cgroup/freezer.go [hunk 3]
```diff
@@ -187,35 +213,37 @@ func freezeSnapProcessesImplV2(snapName string) error {
...
-func thawSnapProcessesV2(snapName string, skipErrs bool) error {
+func thawSnapProcessesV2(snapName string, skipError func(error) bool) error {
 	thawOne := func(dir string) error {
 		fname := filepath.Join(dir, "cgroup.freeze")
-		if err := ioutil.WriteFile(fname, []byte("0"), 0644); err != nil && os.IsNotExist(err) {
+		if err := writeFile(fname, []byte("0"), 0644); err != nil && os.IsNotExist(err) {
 			//  the group may be gone already
 			return nil
-		} else if err != nil && !skipErrs {
+		} else if err != nil && (skipError == nil || !skipError(err)) {
```
Comment (stolowski) [inline sandbox/cgroup/freezer.go:236]: AFAICT we don't test the case where skipError == nil, would be nice to have one.

**Review by stolowski — APPROVED** [review]
LGTM, three minor comments.

File: sandbox/cgroup/freezer.go [hunk 4]
```diff
@@ -140,13 +145,29 @@ func applyToSnap(snapName string, action func(groupName string) error, skipErrs
...
+		if err := action(name); err != nil && (skipError == nil || !skipError(err)) {
 			return err
 		}
 		return filepath.SkipDir
 	})
 }
 
+// writeFile can be used as a drop-in replacement for ioutil.WriteFile, but does
+// not create a file when does not exist
+func writeFile(where string, data []byte, mode os.FileMode) error {
```
Comment (anonymouse64) [inline sandbox/cgroup/freezer.go:157]:
```suggestion
func writeExistingFile(where string, data []byte, mode os.FileMode) error {
```

File: sandbox/cgroup/freezer.go (hunk 1 above)
Comment (anonymouse64) [reply sandbox/cgroup/freezer.go:168]: I think this is more readable too

File: sandbox/cgroup/freezer.go [hunk 5]
```diff
@@ -187,35 +213,37 @@ func freezeSnapProcessesImplV2(snapName string) error {
 		}
 		return fmt.Errorf("cannot freeze processes of snap %q in group %v", snapName, filepath.Base(dir))
 	}
-	const skipFreezeErrs = false
-	err = applyToSnap(snapName, freezeOne, skipFreezeErrs)
+	// freeze skipping ENOENT errors
```
Comment (anonymouse64) [inline sandbox/cgroup/freezer.go:216]:
```suggestion
	// freeze, skipping ENOENT errors
```

File: sandbox/cgroup/freezer.go (hunk 3 above)
Comment (anonymouse64) [reply sandbox/cgroup/freezer.go:236]: that or just always assume it is not nil

**Review by anonymouse64 — APPROVED** [review]
lgtm

File: sandbox/cgroup/freezer.go (hunk 1 above)
Comment (bboozzoo) [reply sandbox/cgroup/freezer.go:168]: done

File: sandbox/cgroup/freezer.go (hunk 3 above)
Comment (bboozzoo) [reply sandbox/cgroup/freezer.go:236]: > that or just always assume it is not nil

And that's probably the easiest case

Comment (bboozzoo) [reply sandbox/cgroup/freezer.go]: I've extended the unit test to at least cover applySnap()

**Pull request merged**